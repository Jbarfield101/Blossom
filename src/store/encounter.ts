import { create } from 'zustand';
import { persist } from 'zustand/middleware';
import { Encounter, Participant } from '../dnd/encounters';

interface EncounterState {
  encounter: Encounter | null;
  startEncounter: (participants: Participant[]) => void;
  nextTurn: () => void;
  endEncounter: () => void;
}

export const useEncounterStore = create<EncounterState>()(
  persist(
    (set) => ({
      encounter: null,
      startEncounter: (participants) =>
        set({ encounter: new Encounter(participants) }),
      nextTurn: () =>
        set((state) =>
          state.encounter ? { encounter: state.encounter.advance() } : state
        ),
      endEncounter: () => set({ encounter: null }),
    }),
    { name: 'encounter-store' }
  )
);

export type { EncounterState };
