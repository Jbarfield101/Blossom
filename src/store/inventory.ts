import { create } from 'zustand';
import { persist } from 'zustand/middleware';
import type { Npc } from '../dnd/schemas/npc';
import type { Item } from '../dnd/schemas/item';

export interface WorldItem extends Item {
  npcIds: string[];
}

interface InventoryState {
  items: Record<string, WorldItem>;
  scanNPCs: (npcs: Npc[]) => void;
}

export const useInventory = create<InventoryState>()(
  persist(
    (set) => ({
      items: {},
      scanNPCs: (npcs) =>
        set((state) => {
          const items = { ...state.items };
          const nameToId: Record<string, string> = {};
          // reset npc references
          for (const [id, item] of Object.entries(items)) {
            nameToId[item.name] = id;
            item.npcIds = [];
          }
          for (const npc of npcs) {
            for (const name of npc.inventory || []) {
              let itemId = nameToId[name];
              if (!itemId) {
                itemId = crypto.randomUUID();
                nameToId[name] = itemId;
                items[itemId] = {
                  id: itemId,
                  name,
                  value: 0,
                  description: '',
                  tags: [],
                  npcIds: [],
                };
              }
              const entry = items[itemId];
              if (!entry.npcIds.includes(npc.id)) {
                entry.npcIds.push(npc.id);
              }
            }
          }
          return { items };
        }),
    }),
    { name: 'inventory-store' }
  )
);

export type { InventoryState };
