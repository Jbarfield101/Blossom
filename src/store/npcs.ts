import { create } from 'zustand';
import { persist } from 'zustand/middleware';
import { invoke } from '@tauri-apps/api/core';

export interface NPC {
  id: string;
  name: string;
  race: string;
  class: string;
  role?: string;
  cr?: number;
  locale?: string;
  tags?: string[];
  personality: string;
  background: string;
  appearance: string;
  portrait?: string;
  hooks?: string[];
  quirks?: string[];
  secrets?: string[];
  stats?: Record<string, string | number>;
  skills?: Record<string, string | number>;
}

interface NPCState {
  npcs: NPC[];
  addNPC: (npc: Omit<NPC, 'id'>) => void;
  removeNPC: (id: string) => void;
  loadNPCs: () => Promise<void>;
}

export const useNPCs = create<NPCState>()(
  persist(
    (set) => ({
      npcs: [],
      addNPC: (npc) =>
        set((state) => ({
          npcs: [...state.npcs, { id: crypto.randomUUID(), ...npc }],
        })),
      removeNPC: (id) =>
        set((state) => ({ npcs: state.npcs.filter((npc) => npc.id !== id) })),
      loadNPCs: async () => {
        const npcs = await invoke<NPC[]>('list_npcs');
        set({ npcs });
      },
    }),
    { name: 'npc-store' }
  )
);
