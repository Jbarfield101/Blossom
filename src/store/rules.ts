import { create } from 'zustand';
import { persist } from 'zustand/middleware';
import { invoke } from '@tauri-apps/api/core';
import type { RuleData } from '../features/dnd/types';

interface RuleState {
  rules: RuleData[];
  loadRules: () => Promise<void>;
  addRule: (rule: Omit<RuleData, 'id'> & { id?: string }) => Promise<void>;
  removeRule: (id: string) => void;
}

export const useRules = create<RuleState>()(
  persist(
    (set) => ({
      rules: [],
      loadRules: async () => {
        const rules = await invoke<RuleData[]>('list_rules');
        set({ rules });
      },
      addRule: async (rule) => {
        const withId: RuleData = { id: rule.id ?? crypto.randomUUID(), ...rule } as RuleData;
        await invoke('save_rule', { rule: withId });
        set((state) => {
          const exists = state.rules.some((r) => r.id === withId.id);
          const rules = exists
            ? state.rules.map((r) => (r.id === withId.id ? withId : r))
            : [...state.rules, withId];
          return { rules };
        });
      },
      removeRule: (id) =>
        set((state) => ({ rules: state.rules.filter((r) => r.id !== id) })),
    }),
    { name: 'rule-store' }
  )
);

export type { RuleData as Rule };
