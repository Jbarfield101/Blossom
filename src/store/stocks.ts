import { create } from 'zustand';
import { invoke } from '@tauri-apps/api/core';

const SYMBOL_MAP: Record<string, string> = {
  BTC: 'BTC-USD',
  ETH: 'ETH-USD',
};

interface Quote {
  price: number;
  changePercent: number;
  history: number[];
  marketStatus: string;
  lastFetched: number;
  error?: string;
}

interface NewsArticle {
  title: string;
  link: string;
  timestamp: number;
  summary: string;
}

interface StockState {
  quotes: Record<string, Quote>;
  pollers: Record<string, ReturnType<typeof setInterval>>;
  symbols: string[];
  news: Record<string, { articles: NewsArticle[]; lastFetched: number }>;
  fetchQuote: (symbol: string) => Promise<number>;
  startPolling: (symbol: string, interval?: number) => void;
  stopPolling: (symbol: string) => void;
  forecast: (symbol: string) => Promise<string>;
  fetchNews: (symbol: string) => Promise<NewsArticle[]>;
  addStock: (symbol: string) => void;
  removeStock: (symbol: string) => void;
}

export const useStocks = create<StockState>((set, get) => ({
  quotes: {},
  pollers: {},
  symbols: [],
  news: {},
  fetchQuote: async (symbol) => {
    const sym = symbol.toUpperCase();
    const fetchSym = SYMBOL_MAP[sym] ?? sym;
    try {
      const bundle = await invoke<{
        quotes: { price: number; change_percent: number; status: string }[];
        series: { points: { ts: number; close: number }[] }[];
        market: { phase: string };
        stale: boolean;
      }>('stocks_fetch', { tickers: [fetchSym], range: '1d' });
      const quote = bundle.quotes[0];
      const series = bundle.series[0];
      const history = series?.points?.map((p) => p.close) ?? [];
      const market_status = bundle.market.phase;
      set((state) => ({
        quotes: {
          ...state.quotes,
          [sym]: {
            price: quote?.price ?? 0,
            changePercent: quote?.change_percent ?? 0,
            history,
            marketStatus: market_status,
            lastFetched: Date.now(),
            error: undefined,
          },
        },
      }));
      return quote?.price ?? NaN;
    } catch (e) {
      const message = e instanceof Error ? e.message : String(e);
      set((state) => {
        const prev =
          state.quotes[sym] ?? {
            price: NaN,
            changePercent: NaN,
            history: [],
            marketStatus: '',
            lastFetched: Date.now(),
          };
        return {
          quotes: {
            ...state.quotes,
            [sym]: { ...prev, error: message, lastFetched: Date.now() },
          },
        };
      });
      return NaN;
    }
  },
  fetchNews: async (symbol) => {
    const sym = symbol.toUpperCase();
    const cached = get().news[sym];
    if (cached && Date.now() - cached.lastFetched < 15 * 60 * 1000) {
      return cached.articles;
    }
    try {
      const articles = await invoke<NewsArticle[]>('fetch_stock_news', {
        symbol: sym,
      });
      set((state) => ({
        news: { ...state.news, [sym]: { articles, lastFetched: Date.now() } },
      }));
      return articles;
    } catch {
      return [];
    }
  },
  startPolling: (symbol, interval = 15000) => {
    const sym = symbol.toUpperCase();
    const { pollers } = get();
    if (pollers[sym]) return;
    const id = setInterval(() => {
      get().fetchQuote(sym);
    }, interval);
    set((state) => ({ pollers: { ...state.pollers, [sym]: id } }));
    get().fetchQuote(sym);
  },
  stopPolling: (symbol) => {
    const sym = symbol.toUpperCase();
    const { pollers } = get();
    const id = pollers[sym];
    if (id) clearInterval(id);
    set((state) => {
      const { [sym]: _, ...rest } = state.pollers;
      return { pollers: rest };
    });
  },
  forecast: async (symbol) => {
    const sym = symbol.toUpperCase();
    try {
      return await invoke<string>('stock_forecast', { symbol: sym });
    } catch {
      return 'Forecast currently unavailable.';
    }
  },
  addStock: (symbol) => {
    const sym = symbol.toUpperCase();
    set((state) =>
      state.symbols.includes(sym)
        ? state
        : { symbols: [...state.symbols, sym] }
    );
    get().startPolling(sym);
  },
  removeStock: (symbol) => {
    const sym = symbol.toUpperCase();
    set((state) => ({ symbols: state.symbols.filter((s) => s !== sym) }));
    get().stopPolling(sym);
  },
}));

export type { Quote, StockState, NewsArticle };
