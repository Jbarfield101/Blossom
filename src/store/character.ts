import { create } from 'zustand';
import { persist } from 'zustand/middleware';
import type { Character } from '../dnd/characters';
import { useNPCs } from './npcs';

interface CharacterState {
  character: Character | null;
  setCharacter: (char: Character) => void;
  clearCharacter: () => void;
}

export const useCharacter = create<CharacterState>()(
  persist(
    (set) => ({
      character: null,
      setCharacter: (char) => {
        set({ character: char });
        const addNPC = useNPCs.getState().addNPC;
        addNPC({
          name: 'Player Character',
          species: 'Unknown',
          role: char.class,
          alignment: 'Unaligned',
          playerCharacter: true,
          backstory: '',
          location: '',
          hooks: [],
          quirks: [],
          appearance: 'Unknown',
          portrait: 'placeholder.png',
          icon: 'placeholder-icon.png',
          sections: {},
          statblock: {},
          tags: ['player'],
          abilities: char.abilities,
          level: char.level,
          hp: char.hp,
          inventory: char.inventory.map((i) => i.name),
        });
      },
      clearCharacter: () => set({ character: null }),
    }),
    { name: 'character-store' }
  )
);

