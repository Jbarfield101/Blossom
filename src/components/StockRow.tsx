import { memo, useEffect, useState } from "react";
import { Box, Typography, IconButton, ListItem } from "@mui/material";
import DeleteIcon from "@mui/icons-material/Delete";
import Sparkline from "./Sparkline";
import { useStocks } from "../store/stocks";

function StockRow({ symbol }: { symbol: string }) {
  const quote = useStocks((state) => state.quotes[symbol]);
  const removeStock = useStocks((state) => state.removeStock);
  const forecastFn = useStocks((state) => state.forecast);
  const fetchNews = useStocks((state) => state.fetchNews);
  const [forecast, setForecast] = useState<{ shortTerm: string; longTerm: string } | null>(null);
  const [newsSummary, setNewsSummary] = useState("");
  useEffect(() => {
    forecastFn(symbol).then(setForecast);
    fetchNews(symbol).then((articles) => {
      const summary = articles
        .slice(0, 3)
        .map((a) => a.title)
        .join("; ");
      setNewsSummary(summary);
    });
  }, [symbol, forecastFn, fetchNews]);
  const color = quote && quote.changePercent < 0 ? "#ff5252" : "#4caf50";

  return (
    <ListItem
      key={symbol}
      secondaryAction={
        <IconButton edge="end" onClick={() => removeStock(symbol)}>
          <DeleteIcon />
        </IconButton>
      }
    >
      <Box sx={{ display: "flex", flexDirection: "column", gap: 1, width: "100%" }}>
        <Box sx={{ display: "flex", alignItems: "center", gap: 2 }}>
          <Typography sx={{ width: 80 }}>{symbol}</Typography>
          {quote ? (
            quote.error ? (
              <Typography sx={{ width: 120, color: "#ff5252" }}>
                {quote.error}
              </Typography>
            ) : (
              <>
                <Typography sx={{ width: 120, color }}>
                  {quote.price.toFixed(2)} ({quote.changePercent.toFixed(2)}%)
                </Typography>
                <Typography sx={{ width: 80 }}>{quote.marketStatus}</Typography>
                <Sparkline data={quote.history} color={color} />
              </>
            )
          ) : (
            <Typography sx={{ width: 120 }}>...</Typography>
          )}
        </Box>
        {forecast && (
          <Box sx={{ display: "flex", flexDirection: "column", gap: 0.5 }}>
            <Typography variant="body2">Short: {forecast.shortTerm}</Typography>
            <Typography variant="body2">Long: {forecast.longTerm}</Typography>
          </Box>
        )}
        {newsSummary && (
          <Typography variant="body2" sx={{ mt: 0.5 }}>
            {newsSummary}
          </Typography>
        )}
      </Box>
    </ListItem>
  );
}

export default memo(StockRow);
