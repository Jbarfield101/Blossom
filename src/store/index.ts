import { saveState, loadState } from "../utils/persist";
import { useNPCs } from "./npcs";
import { useWorlds } from "./worlds";
import { useCharacter } from "./character";
import { useEncounterStore } from "./encounter";
import { useWarTableStore } from "./warTable";
import { useTabletopStore } from "./tabletop";
import { useVoices } from "./voices";
import { useInventory } from "./inventory";

// Export individual stores for existing imports elsewhere
export {
  useNPCs,
  useWorlds,
  useCharacter,
  useEncounterStore,
  useWarTableStore,
  useTabletopStore,
  useVoices,
  useInventory,
};

const slices = {
  npcs: useNPCs,
  worlds: useWorlds,
  character: useCharacter,
  encounter: useEncounterStore,
  warTable: useWarTableStore,
  tabletop: useTabletopStore,
  inventory: useInventory,
};

type StoreState<T> = {
  [K in keyof T]: T[K] extends { getState: () => infer S }
    ? { [P in keyof S as S[P] extends Function ? never : P]: S[P] }
    : never;
};

export type CampaignState = StoreState<typeof slices>;

export async function saveCampaign(
  key = "campaign",
  backendUrl?: string
) {
  const data: Record<string, unknown> = {};
  for (const [name, store] of Object.entries(slices)) {
    const state = store.getState();
    data[name] = JSON.parse(JSON.stringify(state));
  }
  await saveState(key, data, backendUrl);
}

export async function loadCampaign(
  key = "campaign",
  backendUrl?: string
) {
  const data = await loadState<CampaignState>(key, backendUrl);
  if (!data) return;
  for (const [name, store] of Object.entries(slices)) {
    const slice = (data as any)[name];
    if (slice) store.setState(slice, true);
  }
}

