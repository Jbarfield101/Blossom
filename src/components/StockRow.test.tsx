import { describe, it, expect, beforeEach, afterEach, vi } from 'vitest';
import { render, cleanup } from '@testing-library/react';
import StockRow from './StockRow';
import { useStocks } from '../store/stocks';

const baseQuote = {
  price: 123,
  changePercent: 1,
  history: [1, 2, 3],
  marketStatus: 'OPEN',
  volume: 1000,
  lastFetched: Date.now(),
};

describe('StockRow', () => {
  beforeEach(() => {
    useStocks.setState({
      quotes: {},
      ws: {},
      removeStock: vi.fn(),
      forecast: vi.fn().mockResolvedValue({ shortTerm: 'st', longTerm: 'lt' }),
      fetchNews: vi.fn().mockResolvedValue([]),
    } as any);
  });

  afterEach(() => {
    useStocks.setState({ quotes: {}, ws: {}, forecast: undefined, fetchNews: undefined } as any);
    cleanup();
  });

  it('renders quote data', async () => {
    useStocks.setState({ quotes: { AAPL: baseQuote } });
    const { asFragment } = render(
      <StockRow
        symbol="AAPL"
        metrics={{ price: true, change: true, volume: true, trend: true }}
      />
    );
    await Promise.resolve();
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders error message when present', () => {
    useStocks.setState({ quotes: { AAPL: { ...baseQuote, error: 'fail' } } });
    const { getByText } = render(
      <StockRow
        symbol="AAPL"
        metrics={{ price: true, change: true, volume: true, trend: true }}
      />
    );
    expect(getByText('fail')).toBeInTheDocument();
  });
});
