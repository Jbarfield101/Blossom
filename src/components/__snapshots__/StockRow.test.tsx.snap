// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`StockRow > renders quote data 1`] = `
<DocumentFragment>
  <li
    class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1p4qrhw-MuiListItem-root"
  >
    <div
      class="MuiBox-root css-1lq4uud"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-1019lio-MuiTypography-root"
      >
        AAPL
      </p>
      <p
        class="MuiTypography-root MuiTypography-body1 css-tsexvv-MuiTypography-root"
      >
        123.00 (1.00%)
      </p>
      <p
        class="MuiTypography-root MuiTypography-body1 css-1019lio-MuiTypography-root"
      >
        OPEN
      </p>
      <svg
        class="MuiBox-root css-1bayzp8"
      >
        <polyline
          fill="none"
          points="0,30 50,15 100,0"
          stroke="#4caf50"
          stroke-width="2"
        />
      </svg>
    </div>
    <div
      class="MuiListItemSecondaryAction-root css-zezwi5-MuiListItemSecondaryAction-root"
    >
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1ysp02-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
          data-testid="DeleteIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6zM19 4h-3.5l-1-1h-5l-1 1H5v2h14z"
          />
        </svg>
      </button>
    </div>
  </li>
</DocumentFragment>
`;
